Using clang-tidy to correct legacy namings

TotoGaz opened the pull request:
Naming conventions were not respected everywhere.
This patch was mostly generated with `clang-tidy`.
Some manual corrections though.

Also modifies submodule through https://github.com/GEOSX/GEOSX_PTP/pull/22

corbett5 commented:
Wow this is awesome! Obviously I didn't look at all of the changes though. Maybe you should add the script to the `scripts` repo, or at least I'd like to know how you did it haha.

joshua-white commented on src/coreComponents/constitutive/fluid/MultiPhaseMultiComponentFluid.cpp:
```diff
@@ -84,28 +84,28 @@ MultiPhaseMultiComponentFluid::deliverClone( string const & name,
   GEOSX_ERROR_IF( m_phasePVTParaFiles.size() != NP, "The number of phasePVTParaFiles is not the same as the number of phases!" );
 
-  CreatePVTModels();
+  createPvtModels();
```
I'm okay with `Pvt` as I think we want to maximize the automation, but I thought I'd call it out for others.

TotoGaz replied:
done

joshua-white commented on src/coreComponents/constitutive/fluid/PVTFunctions/FenghourCO2ViscosityFunction.cpp:
```diff
@@ -96,15 +96,15 @@ void FenghourCO2ViscosityFunction::MakeTable( string_array const & inputPara )
   real64_array2d densities( nP, nT );
 
-  SpanWagnerCO2DensityFunction::CalculateCO2Density( pressures, temperatures, densities );
+  SpanWagnerCO2DensityFunction::calculateCO2Density( pressures, temperatures, densities );
```
I am guessing you had to hand correct `CO2` from `Co2`?

TotoGaz replied:
Actually I did not think I did 🤷‍♂️

joshua-white commented on src/coreComponents/finiteVolume/mimeticInnerProducts/BdVLMInnerProduct.hpp:
```diff
@@ -127,7 +127,7 @@ BdVLMInnerProduct::Compute( arrayView2d< real64 const, nodes::REFERENCE_POSITION
     // TODO: see what it would take to bring the (harmonically averaged) two-point trans here
     real64 diagEntry = 0.0;
-    MimeticInnerProductHelpers::ComputeInvTPFATransWithMultiplier< NF >( elemPerm,
+    MimeticInnerProductHelpers::computeInvTpfaTransWithMultiplier< NF >( elemPerm,
```
`Tpfa` is another one like `Pvt`, but I suggest we leave as it.  It is still very legible.

TotoGaz replied:
I've put back `TPFA` while surprisingly `Tpfa` is more readable imho. (because you know when to stop reading the letters and start again standard reaging. Same for `PVT`

joshua-white commented on src/coreComponents/linearAlgebra/DofManager.cpp:
```diff
@@ -1305,7 +1305,7 @@ void vectorToFieldImpl( LOCAL_VECTOR const localVector,
   GEOSX_ASSERT( wrapper != nullptr );
 
-  rtTypes::ApplyArrayTypeLambda2( rtTypes::typeID( std::type_index( wrapper->get_typeid() ) ),
+  rtTypes::applyArrayTypeLambda2( rtTypes::typeID( std::type_index( wrapper->getTypeid() ) ),
```
Perhaps we need `getTypeId()`

TotoGaz replied:
Done

joshua-white reviewed (commented):
I agree this is totally awesome.  Some things aren't perfect, but I'd rather have 99% of the code consistent.  Out of curiosity, how much hand cleanup was required?  How regularly do you think we could run a cleanup PR to catch `BadFormatting()`?  Could we do it once a year?

joshua-white commented on src/coreComponents/constitutive/capillaryPressure/CapillaryPressureBase.hpp:
```diff
@@ -124,7 +124,7 @@ class CapillaryPressureBase : public ConstitutiveBase
 
   arrayView3d< real64 const > phaseCapPressure() const { return m_phaseCapPressure; }
-  arrayView4d< real64 const > dPhaseCapPressure_dPhaseVolFraction() const { return m_dPhaseCapPressure_dPhaseVolFrac; }
+  arrayView4d< real64 const > dPhaseCapPressureDPhaseVolFraction() const { return m_dPhaseCapPressure_dPhaseVolFrac; }
```
New derivative style, fine by me though.

TotoGaz replied:
I have forgotten this correction.
The underscore is back.

TotoGaz commented:
> I agree this is totally awesome. Some things aren't perfect, but I'd rather have 99% of the code consistent.

I went through all the files and I could not spot any mistake (unified diff is great here).
Surely some modifications may be change in the future with far far ... far far smaller PRs.
Personally I would merge it ASAP, and correct after if needed.

_Probably there will be something to do with python bindings._
 
> Out of curiosity, how much hand cleanup was required? 

First commit is the result of `clang-tidy`. The rest is corrections.
`clang-tidy` definitely got lost in deep template stuff. And also because I had to trick him for submodules.
But it made the ~95% for us.

> How regularly do you think we could run a cleanup PR to catch `BadFormatting()`? Could we do it once a year?

Today was huge because we had so much legacy.
This won't happen anymore and our PRs will protect us.
Yes, once a year could be enough: people won't deliberately disrespect the naming conventions.
Our we could put a robot on our CI/CD, but we are not 100% consistent.
So that won't be easy.

Also, checking is quite easy. Correcting is another story.

TotoGaz commented:
`travis-ci` is ok (just the submodule check is red) so we can merge if there is no major concern.

joshua-white commented:
I am okay with merging.  @rrsettgast?

sytuannguyen commented on src/coreComponents/dataRepository/Group.hpp:
```diff
@@ -1471,7 +1471,7 @@ class Group
    */
-  virtual void InitializePostInitialConditions_PostSubGroups( Group * const group )
+  virtual void initializePostInitialConditionsPostSubGroups( Group * const group )
   {
```
This seem to be the longest function name, can we give it another more appropriate name?

TotoGaz replied:
Yes this is true, but I could only find this name like 4 times in the whole codebase.
Also the wording is quite comprehensive and every word has its importance.
Se could transform it into `initPostInitialCondsPostSubGrps`, but is it really worth it for so unfrequent use?

sytuannguyen reviewed (commented):
Awsome! Can this tool used to clean also the "const rule"? 
For see both "std::string const &" and "const std::string &", I see also "string const &" and "const string &".

Could you check also to homogenize the indexing style? You can see that both [i] and [ i ] exist. Similarly for the parenthesis, you can see that the space some time is added some time not: e.g. ( blabla ), (blabla ), ( blabla) and (blabla).

TotoGaz commented:
> Awsome!

Thx
 
> Can this tool used to clean also the "const rule"?
> For see both "std::string const &" and "const std::string &", I see also "string const &" and "const string &".

I could not find anything in https://clang.llvm.org/extra/clang-tidy/checks/list.html and nothing through google (shallow searching though)

> Could you check also to homogenize the indexing style? You can see that both [i] and [ i ] exist. Similarly for the parenthesis and brackets, you can see that the space some time is added some time not: e.g. ( blabla ), (blabla ), ( blabla), (blabla), {blabla}, { blabla }. Should we add also a rule for this?

`clang-tidy` is more powerful than just formatting (note that `clang` family has `clang-format` for this), since it analyzes and modifies the code itself, not just spaces.
I'm quite sure that `uncrustify` allows that. Maybe some files were not controlled. Maybe some further action?
But first a rule needs to be defined, this would be the longest.

rrsettgast commented:
> 
> > Could you check also to homogenize the indexing style? You can see that both [i] and [ i ] exist. Similarly for the parenthesis and brackets, you can see that the space some time is added some time not: e.g. ( blabla ), (blabla ), ( blabla), (blabla), {blabla}, { blabla }. Should we add also a rule for this?
> 
> `clang-tidy` is more powerful than just formatting (note that `clang` family has `clang-format` for this), since it analyzes and modifies the code itself, not just spaces.
> I'm quite sure that `uncrustify` allows that. Maybe some files were not controlled. Maybe some further action?
> But first a rule needs to be defined, this would be the longest.

We certainly can control spaces with uncrustify, and at some point we had a control in place for this. However, it gets a little hard to control with nested indexing. For instance something like `variable[mapA[a]]` is a little hard to read while `variable[ mapA[ a ] ]` is easier...but then for long equations we are taking a lot of columns on spaces...Classic readability vs compactness tradeoff. So we allow both. We don't have control over when to apply formatting in a non-nested case, and when not to apply formatting in a nested case.

rrsettgast commented on src/coreComponents/codingUtilities/static_if.hpp:
```diff
@@ -28,14 +28,14 @@ template<bool CONDITION>
 {
   template<typename LAMBDA_BODY>
-  constexpr inline static void if_function(LAMBDA_BODY&&) { }
+  constexpr inline static void ifFunction(LAMBDA_BODY&&) { }
```
@corbett5 @klevzoff @wrtobin Are we still using these?

TotoGaz replied:
@rrsettgast so I'm starting the merge! (submodules come first)
Anything can be modified in a future PR if needed.

TotoGaz commented:
> We certainly can control spaces with uncrustify, and at some point we had a control in place for this. However, it gets a little hard to control with nested indexing. For instance something like `variable[mapA[a]]` is a little hard to read while `variable[ mapA[ a ] ]` is easier...but then for long equations we are taking a lot of columns on spaces...Classic readability vs compactness tradeoff. So we allow both. We don't have control over when to apply formatting in a non-nested case, and when not to apply formatting in a nested case.

Byt the way, we may be using something like https://clang.llvm.org/docs/ClangFormatStyleOptions.html#disabling-formatting-on-a-piece-of-code (`clang-format off/on`) or `/* *INDENT-ON/OFF* */` for `uncrustify`.
So we would have 
```c++
/* *INDENT-OFF* */
a[ i + 1 ] + b [ i     ]
a[ i     ] + b [ i + 1 ]
/* *INDENT-ON* */
```

TotoGaz commented:
> Maybe you should add the script to the `scripts` repo, or at least I'd like to know how you did it haha.

So actually this is how I did:

I did create the `.clang-tidy` file at the root of sources.
It contains the configuration elements.
Mine was like
```
Checks: '-*,readability-identifier-naming'
CheckOptions:
- { key: readability-identifier-naming.MethodCase, value: camelBack }
HeaderFilterRegex: coreComponents/codingUtilities|coreComponents/common|coreComponents/constitutive|coreComponents/dataRepository|coreComponents/fileIO|coreComponents/finiteElement|coreComponents/finiteVolume|coreComponents/linearAlgebra|coreComponents/managers|coreComponents/math|coreComponents/mesh|coreComponents/meshUtilities|coreComponents/mpiCommunications|coreComponents/physicsSolvers|coreComponents/rajaInterface|coreComponents/virtualElement 
```
This strange `HeaderFilterRegex` is because I have to define the `c++` headers I want to be considered.
Since `LvArray` is an external component in the `coreComponents` folder, I had to filter.
My attempts at coding some “_does not match_” regexp did not succeed inside `clang-tidy`…

I also added the `-DCMAKE_EXPORT_COMPILE_COMMANDS=ON` because `clang-tidy` uses this json output to know how it must generate the AST (I think).

Then I tricked to filter elements I did not want to analyse (I removed `LvArray`, `externalComponents`, `blt`). Maybe this was improper…

Then run it with `run-clang-tidy -p cmake-debug-build -export-fixes fixes.yaml`.
I had to `sed` `fixes.yaml` because I do not know why but relative paths (`FilePath: ../coreComponents/…`) were not resolved appropriately.
Then apply with `clang-apply-replacements .`.

Last, try to recompile, see where it breaks, fix…
